### PR TITLE
Closed adding initial support for CSV batch ingest. dams5-cc-pilot #47

### PR DESCRIPTION
**Pull request by lsitu**
Related to https://github.com/ucsdlib/dams5-cc-pilot/issues/47

### CODE REVIEW

**Comment by lsitu** [comment]
@ucsdlib/developers Please review, comments ... Thanks

File: app/controllers/concerns/csv_imports_controller_behavior.rb [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+          redirect_to sufia.dashboard_works_path
+        end
+      end
+
+      def create_update_job
+        log = Sufia::BatchCreateOperation.create!(user: current_user,
+                                           operation_type: "CSV Import")
+
+        uploaded_files = params.fetch(:uploaded_files, [])
+        selected_files = params.fetch(:selected_files, {}).values
```
Comment (mcritchlow) [inline app/controllers/concerns/csv_imports_controller_behavior.rb]: If `uploaded_files` or `selected_files` hit the fallback case for `fetch` (returning [] or {}), then should we use that as a guard class and return? It looks like it's calling the `perform_later` method on the job regardless of there being files or not. Or, are files optional and this is desired behavior?

File: app/jobs/csv_import_job.rb [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        f = Sufia::UploadedFile.find(id).file if !id.blank?
+        uploaded_files_map[File.basename(f.current_path)] = id
+      end
+
+      selected_files.each do |f|
+        remote_files_map[f[:file_name]] = f
+      end
+
+      model = model_to_create(attributes) #::ObjectResource #
+      count = 0
```
Comment (mcritchlow) [inline app/jobs/csv_import_job.rb]: what is the purpose of `count`? I only see it being incremented at the end of the each block, but not used.

File: app/jobs/csv_import_job.rb [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+      parser = CSVParser.new(csv_source)
+      parser.each do |attrs|
+        splict_attributes (attrs)
```
Comment (mcritchlow) [inline app/jobs/csv_import_job.rb]: is this supposed to be `splice_attributes`?

File: app/jobs/csv_import_job.rb [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        auth_arr = attrs.delete(:Language)
+        attrs = attrs.merge(language: auth_arr)
+
+        # Resource Type field
+        auth_arr = attrs.delete(:Type)
+        attrs = attrs.merge(resource_type: auth_arr)
+
+        matched_files = []
+        remote_files = []
+        if attrs.include?(:file_1)
```
Comment (mcritchlow) [inline app/jobs/csv_import_job.rb]: `:file_1` and `:file_2` seem like fairly arbitrary names. Are there more meaningful names we could use? Is one referring to a master file, for example?

File: app/jobs/csv_import_job.rb [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        matched_files = []
+        remote_files = []
+        if attrs.include?(:file_1) 
+          file_name = attrs[:file_1].first
+          matched_files << uploaded_files_map[attrs[:file_1].first] if uploaded_files_map.key?(file_name)
+          remote_files << remote_files_map[attrs[:file_1].first] if remote_files_map.key?(file_name)
+          attrs.delete(:file_1)
+        end
+
+        if attrs.include?(:file_2)
```
Comment (mcritchlow) [inline app/jobs/csv_import_job.rb]: These two if blocks, for `:file_1` and `:file_2` are exactly the same. Perhaps there's a method to extract here?

File: app/jobs/csv_import_job.rb [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+    # Override this method if you have a different rubric for choosing the model
+    # @param [Hash] attributes
+    # @return String the model to create
+    def model_to_create(attributes)
+      Sufia.config.model_to_create.call(attributes)
+    end
+
+    def authority_hash (attrs, key, model)
+      authority_arr = []
```
Comment (mcritchlow) [inline app/jobs/csv_import_job.rb]: you might be able to use either `map` directly on `attrs` or `each_with_object` here to build the authority array.

File: app/jobs/csv_import_job.rb [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+      if(attrs.key?(key))
+        attrs.delete(key).each do |label|
+          authority_arr << AuthoritiesService.find_or_create(model, label)
+        end
+      end
+      authority_arr
+    end
+
+    def splict_attributes (attrs)
```
Comment (mcritchlow) [inline app/jobs/csv_import_job.rb]: same name question, should this be `splice_attributes`?

File: app/controllers/concerns/csv_imports_controller_behavior.rb (hunk 1 above)
Comment (lsitu) [inline app/controllers/concerns/csv_imports_controller_behavior.rb]: @mcritchlow Yes, the source files are optional for our ingest and we can choose from either uploaded_files (from local disk) or selected_files (from server side or other share drives), or even both as what've configured, and [] or {} are guard classes when calling the the perform_later method on the job.

File: app/jobs/csv_import_job.rb (hunk 2 above)
Comment (lsitu) [inline app/jobs/csv_import_job.rb]: Yes, the counter hasn't been used yet and we can remove it for now.

File: app/jobs/csv_import_job.rb (hunk 3 above)
Comment (lsitu) [inline app/jobs/csv_import_job.rb]: Yes, it's a typo for split_attributes, and splice_attributes makes sense too.

Comment (mcritchlow) [inline app/jobs/csv_import_job.rb]: I see. Yep, split probably makes the most sense. 👍

File: app/jobs/csv_import_job.rb (hunk 4 above)
Comment (lsitu) [inline app/jobs/csv_import_job.rb]: Yes, `:file_1` and and `:file_2` are arbitrary names from the CSV template header (borrowed from the dams4 Excel Input Stream) and we only can change it when the new CSV template is changed.

File: app/jobs/csv_import_job.rb (hunk 5 above)
Comment (lsitu) [inline app/jobs/csv_import_job.rb]: Sure. We can extract it out.

**Comment by lsitu** [comment]
@mcritchlow I've added commit https://github.com/ucsdlib/horton/pull/10/commits/db3662735228ddf6557949ae3a809bf2b5f100de to address the issues you suggested above. Please let me know if I miss anything. Thanks.

**Comment by hweng** [comment]
:+1:

**Comment by VivianChu** [comment]
👍

**Pull request merged**